Add Admin functionality and refactor dependency injection

aabdoo23 opened the pull request:
Refactored AuthController and SchedsForumsDbContext to use constructor injection. Updated SchedsForums.Application.csproj to include Admin commands. Added AdminController with endpoint for creating admin users, protected by an "Admin" role policy. Introduced CreateAdminCommand, CreateAdminCommandHandler, CreateAdminResponseDTO, and CreateAdminValidator for handling admin creation. Added Admin and Moderator entities inheriting from BaseUser. Updated ServiceCollectionExtensions to register IBaseRepository<Admin> and refactor AddJwtAuthentication to use IOptions<JwtOptions>.

aabdoo23 committed:
Add Admin functionality and refactor dependency injection

Refactored AuthController and SchedsForumsDbContext to use constructor injection. Updated SchedsForums.Application.csproj to include Admin commands. Added AdminController with endpoint for creating admin users, protected by an "Admin" role policy. Introduced CreateAdminCommand, CreateAdminCommandHandler, CreateAdminResponseDTO, and CreateAdminValidator for handling admin creation. Added Admin and Moderator entities inheriting from BaseUser. Updated ServiceCollectionExtensions to register IBaseRepository<Admin> and refactor AddJwtAuthentication to use IOptions<JwtOptions>.

MostafaHgz commented on SchedsForums.Api/Controllers/AdminController.cs:
```diff
@@ -0,0 +1,22 @@
+    {
+        private readonly IMediator _mediator = mediator ?? throw new ArgumentNullException(nameof(mediator));
+
+        [HttpPost("CreateAdmin")]
```
no need to ad action name. just POST on admin means it's for creating an admin

MostafaHgz commented on SchedsForums.Application/Commands/Admins/Create/CreateAdminCommand.cs:
```diff
@@ -0,0 +1,12 @@
+        public string FullName { get; set; }
+        public string UserName { get; set; }
+        public string Email { get; set; }
+        public string Password { get; set; }
```
why does an admin create another admin with its password

MostafaHgz commented on SchedsForums.Application/Commands/Admins/Create/CreateAdminValidator.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        public CreateAdminValidator()
+        {
+            RuleFor(student => student.FullName)
```
this is the same validator in the user.

MostafaHgz commented on SchedsForums.Application/Commands/Admins/Create/CreateAdminCommandHandler.cs:
```diff
@@ -0,0 +1,36 @@
+
+            return new CreateAdminResponseDTO
+            {
+                FullName = admin.Name,
```
you could use the same response as user response dto here

MostafaHgz replied:
also the user response should contain the id

MostafaHgz commented on SchedsForums.Application/Commands/Admins/Create/CreateAdminCommandHandler.cs:
```diff
@@ -0,0 +1,36 @@
+
+            return new CreateAdminResponseDTO
+            {
+                FullName = admin.Name,
```
you could use the same response as user response dto  here

MostafaHgz commented on SchedsForums.Application/Commands/Admins/Create/CreateAdminCommand.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace SchedsForums.Application.Commands.Admins.Create
+{
+    public class CreateAdminCommand : IRequest<CreateAdminResponseDTO>
```
this could extend the user request

aabdoo23 committed:
Refactor user sign-up commands to reduce duplication

Refactored `CreateAdminCommand` and `StudentSignUpCommand` to inherit from a new abstract class `UserSignUpCommand<TResponse>`, which consolidates common properties. Introduced `UserSignUpCommand<TResponse>` in `UserSignUpCommand.cs` under `SchedsForums.Application.Commands.Common`. Removed redundant property declarations and `using MediatR;` directives from the command classes.

MostafaHgz commented on SchedsForums.Domain/Entities/Users/Admin.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public class Admin : BaseUser
+    {
+        //No members are needed for this PR
```
remove comments

MostafaHgz commented on SchedsForums.Infrastructure/Configurations/BaseUserConfiguration.cs:
```diff
@@ -11,7 +11,8 @@ public void Configure(EntityTypeBuilder<BaseUser> builder)
             builder
                 .HasDiscriminator<string>("UserType")
-                .HasValue<Student>("Student");
+                .HasValue<Student>("Student")
```
use nameof(entity)

MostafaHgz commented on SchedsForums.Domain/Entities/Users/Moderator.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Common;
+
+namespace SchedsForums.Domain.Entities.Users
```
moderator is not in this pr yet

MostafaHgz commented on SchedsForums.Application/SchedsForums.Application.csproj:
```diff
@@ -7,13 +7,10 @@
   </PropertyGroup>
 
   <ItemGroup>
```
remove this whole part

aabdoo23 committed:
Some cool DRY refactoring

Refactor DTOs and validators; clean up project files

Refactored `CreateAdminResponseDTO` and `StudentSignUpResponseDTO` to inherit from the new `UserResponseDTO` base class, removing redundant properties. Updated `CreateAdminValidator` and `StudentSignUpValidator` to inherit from the new `UserSignUpValidator` abstract class, consolidating validation rules. Modified `AdminController` to remove specific route from `[HttpPost]` attribute. Cleaned up `SchedsForums.Application.csproj` by removing unnecessary entries. Simplified `Admin` class and removed `Moderator` class. Updated `BaseUserConfiguration` to use `nameof` for discriminator values.

MostafaHgz commented on SchedsForums.Api/Controllers/AdminController.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    [ApiController]
+    [Route("api/[controller]")]
+    [Authorize(Roles = "Admin")]
```
use nameof

aabdoo23 replied:
okey dokey

aabdoo23 committed:
PR done woo

Refactor AdminController authorization

Added `using SchedsForums.Domain.Entities.Users;` directive to reference the `Admin` class. Updated `[Authorize(Roles = "Admin")]` to `[Authorize(Roles = nameof(Admin))]` for better maintainability and to reduce the risk of typos or mismatches.
